dns: refactor cares_wrap internals

jasnell opened the pull request:
This is the first step of a significant refactoring of the DNS subsystem. The first step is to simplify and modernize the code structure, eliminate some duplicate code, and apply some of the more modern patterns that we have implemented elsewhere in the code.

Additional refactorings will come in additional PRs.

Signed-off-by: James M Snell <jasnell@gmail.com>

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

targos commented:
I guess this is now in conflict with https://github.com/nodejs/node/pull/38184 ?

XadillaX commented:
> I guess this is now in conflict with #38184 ?

same question too.

jasnell commented:
No. I've talked to @devsnek about it. That other pr is semver major and makes a number of larger breaking changes. This is semver-patch and is good for everything >= 16.x.

addaleax commented:
@jasnell So, just practically speaking … it’s virtually impossible to review PRs that move large chunks of code around **and** make some minor adjustments in a bunch of places. I’m not sure how to progress here.

targos commented:
I'm like Anna, I don't know how to tackle this review.

jasnell commented:
I'd look at it as a new contribution, largely ignoring the old bits that were changed, focus on the new bits. There's only one test change because one of the internal classes was renamed. There should be absolutely no change in actual behavior.

addaleax commented:
@jasnell I think in that case there will be *a lot* of comments … but yeah, I can do that sometime this week

jasnell commented:
The changes that were made are:

* Class declarations were removed from node_cares.cc and moved into a new node_cares.h
* `node_ares_task` was renamed to `NodeAresTask` and the associated `Hash` and `Equal` impls were moved into `NodeAresTask`.
* `node_ares_task_list` was changed to a `using List` inside `NodeAresTask`.
* `class ChannelWrap :` => `class ChannelWrap final :`
* `class GetAddrInfoReqWrap :` => `class GetAddrInfoReqWrap final :`
* `class GetNameInfoReqWrap :` => `class GetNameInfoReqWrap final :`
* `class QueryWrap` was changed into a `final` template class. Rather than being subclassed for each individual query type, a `using ... = QueryWrap<...>` pattern is used for each type, matching the changes that have been made elsewhere in the codebase (e.g. in `crypto`).
* Added the `struct QueryTraits` template to allow eliminating duplicated code (previously, every query type subclass had nearly identical implementations of `Send()`.
* Eliminated a dead code path (an unused override of `Send()` that was in every query type subclass.
* Introduced the `HostentPointer` smart pointer (using `DeleteFnPtr`) to make cleanup of `hostent*` a bit nicer
* Other small cleanups to match formatting and eliminate unnecessary assignments (e.g. removing `auto context = env->context()` calls and replacing `context` with `env->context()`.

There should be no observable functional changes with the exception of the rename of `node_ares_task_list` to `NodeAresTask::List` in a generated heapdump. All existing dns related tests and benchmarks should continue to operate correctly without modification.

jasnell commented:
@addaleax
> I think in that case there will be a lot of comments

I would expect nothing less ;-)

This particular bit of code has always been far from perfect and I have many more changes in the works to improve it so a detailed review is helpful.

mcollina commented on src/cares_wrap.h:
```diff
@@ -0,0 +1,639 @@
+  /* work after work */
+  dest->h_length = src->h_length;
+  dest->h_addrtype = src->h_addrtype;
+}
```
These are a lot of `memcpy`s. Are all of those needed and could we do with a few less?

addaleax replied:
The cost of the `Malloc` calls here is likely going go *far* outweigh the cost of the `memcpy` calls. If we do optimize, we should try to reduce the number of separate memory allocations (which, luckily, is also much easier to do).

jasnell replied:
While I think we can definitely make perf improvements around this, I'd rather do so in a separate PR. This part of the code in unchanged with this refactor.

mcollina commented on src/cares_wrap.h:
```diff
@@ -0,0 +1,639 @@
+
+  void QueueResponseCallback(int status) {
+    BaseObjectPtr<QueryWrap<Traits>> strong_ref{this};
+    env()->SetImmediate([this, strong_ref](Environment*) {
```
Are we doing some fake-async work? Why the `SetImmediate`?

addaleax replied:
git blame has the answer: 5a948f6f64a538dedce969b21cde454600197ded

tl;dr: if you call `dns.setServers()` from a synchronous callback, the process is going to crash.

mcollina replied:
Amazing work! We should really add a comment here.

jasnell replied:
This is one bit that I've never really been happy with. Rather than introducing the `SetImmediate()` call here, we really ought to block the ability to call `dns.setServers()` from inside a synchronous callback. Calling `setServers()` is unlikely ever to be a hotpath, while the cost of the `SetImmediate()` is incurred on every query. Unfortunately, changing that behavior would be semver-major and is out of scope for this particular PR.

addaleax replied:
I think this is fine, tbh. `SetImmediate()` is a fairly low-overhead operation, compared to the actual I/O.

jasnell replied:
Either way, I don't plan on addressing that in this PR.

mcollina commented on src/cares_wrap.cc:
```diff
@@ -1172,694 +580,622 @@ int ParseSoaReply(Environment* env,
-  QueryCaaWrap(ChannelWrap* channel, Local<Object> req_wrap_obj)
-      : QueryWrap(channel, req_wrap_obj, "resolveCaa") {
+  if (uv_poll_init_socket(channel->env()->event_loop(),
+                          &task->poll_watcher, sock) < 0) {
```
Does this mean we are creating a new socket for every DNS resolution?

addaleax replied:
No. We are creating a `uv_poll_t` here, not a socket.

jasnell replied:
Even the one uv_poll_t per DNS resolution has it's own non-trivial cost. I'd prefer to find a more efficient solution but if @devsnek is successful in getting the transition to `getdns` going and landed, it's likely not worth the effort to refactor this.

mcollina commented on src/cares_wrap.cc:
```diff
@@ -1172,694 +580,622 @@ int ParseSoaReply(Environment* env,
+  if (timeout == 0) timeout = 1;
+  if (timeout < 0 || timeout > 1000) timeout = 1000;
+  uv_timer_start(timer_handle_, AresTimeout, timeout, timeout);
+}
```
Can we move this timer logic in JS? Creating a lot of timers on libuv is somewhat expensive from other analysis.

addaleax replied:
I would not move them unless we can actually prove that that’s worth it.

The reason that we switched the JS timers to not always create libuv timers was the cost of the JS/C++ boundary crossing, not inefficiency of libuv timers, iirc.

mcollina replied:
That's true - however we are always creating these whenever we do a dns lookup, which we do all the time for every new socket created. Those are a lot of timers in addition to everything else.

There are probably better ways to cater for this, possibly deduplicating resolves or other tricks. Closing for now.

jasnell replied:
Similar to the uv_poll_t, it's not likely going to be worth investing the time to refactor that if `getdns` is going to land.

mcollina reviewed (approved):
lgtm

addaleax commented on src/cares_wrap.h:
```diff
@@ -0,0 +1,639 @@
+      struct hostent* host) {
+    UNREACHABLE();
+  }
+};
```
This class seems like an unnecessary indirection? It doesn’t seem to provide anything that we couldn’t put into `Impl` or `QueryWrap` itself

jasnell replied:
Possibly. `ReverseWrap` is the only variant that uses the `hostent` variant of `Parse` and that variant is the only reason why this is added. I can take a pass at refactoring this out. I hadn't already because I was trying to minimize the changes to `QueryWrap`.

jasnell replied:
Ok, the intermediate struct has been removed.

addaleax commented on src/cares_wrap.cc:
```diff
@@ -953,34 +363,34 @@ int ParseTxtReply(Environment* env,
+      ret->Set(env->context(), offset + i, elem).Check();
     } else {
-      ret->Set(context, offset + i, txt_chunk).Check();
+      ret->Set(env->context(), offset + i, txt_chunk).Check();
```
Just going to point out that these changes will all have to be undone if we add support for multi-context builtins

jasnell replied:
Is there any work happening on that right now? And would that also be true for many large chunks of core code? And would that even be relevant here if the `getdns` alternative impl lands? (I have to assume that multi-context builtins will take quite a bit longer to land than the `getdns` impl and that we likely wouldn't backport it to 14 or 16).

addaleax replied:
> Is there any work happening on that right now? 

Not me, at least. 

> And would that also be true for many large chunks of core code? 

Probably, yes.

> And would that even be relevant here if the `getdns` alternative impl lands? (I have to assume that multi-context builtins will take quite a bit longer to land than the `getdns` impl and that we likely wouldn't backport it to 14 or 16).

I mean, it would affect that implementation as well, yes.

jasnell replied:
> I mean, it would affect that implementation as well, yes.

Will worry about that when it becomes an issue then :-)

jasnell commented:
@addaleax ... this is generally ready to go. Did you have any more feedback on it before I land later today?

addaleax commented:
@jasnell No, feel free to merge :+1:

jasnell commented:
Awesome ok. I know there's still a lot in this code that can be improved but with the possible move to getdns in the work, I'm not sure how much more effort we should put into it. I'll merge later on this afternoon.

jasnell commented:
Landed in bc31dc0e0fab
